<a href="https://colab.research.google.com/github/zad-AIworld/Order-processing/blob/main/Add_PRODUCT_ATRRIBUTES.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Adding Product Attributes to the main transaction Data.

Create a PRODUCT_CATEGORY Table using a CSV file from Gemini

The create a PRODUCT_CATEGORY TABLE with P00001 to P0000N naming convention.

Join the Table created above with the order history




In [4]:
import sqlite3

db_path = '/content/Categorized_DB_Enhanced (2).db'

try:
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
    tables = cursor.fetchall()

    print("Tables in the database:")
    for table in tables:
        print(table[0])

except sqlite3.Error as e:
    print(f"Error connecting to the database: {e}")

finally:
    if conn:
        conn.close()

Tables in the database:
order_history
Category_enhanced_orders
Demographics
PRODUCT_CATEGORY


In [3]:
import sqlite3

db_path = '/content/Categorized_DB_Enhanced (2).db'
table_name = 'PRODUCT_CATEGORY'

try:
    conn = sqlite3.connect(db_path)
    # Write the DataFrame to a new table in the database
    product_df.to_sql(table_name, conn, if_exists='replace', index=False)

    print(f"DataFrame successfully written to table '{table_name}' in the database.")

    # Verify by listing tables
    cursor = conn.cursor()
    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
    tables = cursor.fetchall()

    print("\nTables in the database after adding the new table:")
    for table in tables:
        print(table[0])

except sqlite3.Error as e:
    print(f"Error writing DataFrame to database: {e}")

finally:
    if conn:
        conn.close()

DataFrame successfully written to table 'PRODUCT_CATEGORY' in the database.

Tables in the database after adding the new table:
order_history
Category_enhanced_orders
Demographics
PRODUCT_CATEGORY


In [2]:
import pandas as pd

# Read the JSON file into a DataFrame
try:
    product_df = pd.read_json('/content/product_category_gemini.json')
except Exception as e:
    print(f"Error reading the JSON file: {e}")
    product_df = pd.DataFrame() # Create an empty DataFrame in case of error

if not product_df.empty:
    # Get unique values
    product_df = product_df.drop_duplicates()

    # Sort alphabetically (assuming the relevant column for sorting is the first one)
    product_df = product_df.sort_values(by=product_df.columns[0]).reset_index(drop=True)

    # Assign Product_NUM
    product_df['Product_NUM'] = [f'P{i:05d}' for i in range(1, len(product_df) + 1)]

    # Reorder columns to make Product_NUM the first column
    cols = ['Product_NUM'] + [col for col in product_df.columns if col != 'Product_NUM']
    product_df = product_df[cols]

    display(product_df.head())
    print(f"Shape of the product DataFrame: {product_df.shape}")
else:
    print("Product DataFrame is empty or could not be loaded.")

Product_NUM                                       Product_Name  \
0      P00001  10 X AOSANITY FFP2 Atemschutzmaske - Konforme ...   
1      P00002  100W USB C auf Multi USB Kabel [1.2M] Schnelll...   
2      P00003  1m 8K HDMI Kabel 2.1 - 8K@60Hz 4K @ 120Hz DSC ...   
3      P00004  2 Stück Reiseadapter Weltweit auf Deutschland...   
4      P00005  20 Stück Corona Schnelltest Selbsttest - Nase...   

                 Category      Sub_Category     Functional_Use  \
0  Health & Personal Care   Medical Product  Health Protection   
1     Electronics & Photo    Charging Cable    Charging / Data   
2     Electronics & Photo             Cable   Video Connection   
3     Electronics & Photo  Travel Accessory      Power Adapter   
4  Health & Personal Care   Medical Product       Medical Test   

      Physical_Property Age_Group DIY_Product Price_Range  
0               Textile       All          No         N/A  
1  Electronic Component       All          No         N/A  
2  Electronic Component       All          No         N/A  
3       Plastic / Metal     Adult          No         N/A  
4              Test Kit       All          No         N/A

Shape of the product DataFrame: (246, 9)


# Task
Join the `Category_enhanced_orders` and `PRODUCT_CATEGORY` tables from the "Categorized_DB_Enhanced (2).db" database on the `productname` and `product_name` fields respectively, using fuzzy matching for German language values, and save the result to a new table named `Orders_Product_Enhanced` in the same database.

## Load data

### Subtask:
Read the `Category_enhanced_orders` and `PRODUCT_CATEGORY` tables into pandas DataFrames.


**Reasoning**:
Read the specified tables from the database into pandas DataFrames and display their head and shape.



In [5]:
import sqlite3
import pandas as pd

db_path = '/content/Categorized_DB_Enhanced (2).db'

try:
    conn = sqlite3.connect(db_path)

    # Read Category_enhanced_orders table into a DataFrame
    orders_df = pd.read_sql_query("SELECT * FROM Category_enhanced_orders", conn)

    # Read PRODUCT_CATEGORY table into a DataFrame
    products_df = pd.read_sql_query("SELECT * FROM PRODUCT_CATEGORY", conn)

    # Display the first few rows and shape of orders_df
    print("Orders DataFrame:")
    display(orders_df.head())
    print(f"Shape of Orders DataFrame: {orders_df.shape}")

    # Display the first few rows and shape of products_df
    print("\nProducts DataFrame:")
    display(products_df.head())
    print(f"Shape of Products DataFrame: {products_df.shape}")

except sqlite3.Error as e:
    print(f"Error reading tables from database: {e}")

finally:
    if conn:
        conn.close()

Orders DataFrame:


eventDate                  orderDate marketplace  \
0  2025-07-17 14:36:06+00:00  2025-07-17 14:36:06+00:00          DE   
1  2025-07-08 05:06:46+00:00  2025-07-08 05:06:46+00:00          DE   
2  2025-07-04 05:36:24+00:00  2025-07-04 05:36:24+00:00          DE   
3  2025-07-02 19:54:58+00:00  2025-07-02 19:54:58+00:00          DE   
4  2025-06-17 17:33:15+00:00  2025-06-17 17:33:15+00:00          DE   

                                         productName    website  \
0  MELLERUD Schimmel Vernichter Aktivchlor | 1 x ...  Amazon.de   
1  Carson MicroFlip 100x-250x Taschenmikroskop mi...  Amazon.de   
2  Gosknor 2stk Octopus-Kopfhautmassagegerät, Har...  Amazon.de   
3      AJINOMOTO - Monosodium Glutamat, (1 X 200 GR)  Amazon.de   
4  Rainbow Designs Harry Potter Plüsch-Babyrassel...  Amazon.de   

               orderId currencyCode  quantity productCondition  totalOwed  \
0  028-1498008-9740330          EUR         1              New       7.99   
1  028-4603470-0628322          EUR         1              New      24.12   
2  028-5613038-3064361          EUR         1              New       9.59   
3  028-1018659-6139555          EUR         1              New       3.69   
4  028-4961403-1482742          EUR         1              New      14.05   

           Broad Category         Specific Category    USER actual_order_date  \
0          Home & Kitchen         Cleaning Supplies  USER A        2025-07-17   
1             Electronics      Optics & Accessories  USER A        2025-07-08   
2  Beauty & Personal Care          Hair & Skin Care  USER A        2025-07-04   
3  Grocery & Gourmet Food       Cooking Ingredients  USER A        2025-07-02   
4            Toys & Games  Collectibles & Baby Toys  USER A        2025-06-17   

  day_of_the_week  
0        Thursday  
1         Tuesday  
2          Friday  
3       Wednesday  
4         Tuesday

Shape of Orders DataFrame: (266, 15)

Products DataFrame:


Product_NUM                                       Product_Name  \
0      P00001  10 X AOSANITY FFP2 Atemschutzmaske - Konforme ...   
1      P00002  100W USB C auf Multi USB Kabel [1.2M] Schnelll...   
2      P00003  1m 8K HDMI Kabel 2.1 - 8K@60Hz 4K @ 120Hz DSC ...   
3      P00004  2 Stück Reiseadapter Weltweit auf Deutschland...   
4      P00005  20 Stück Corona Schnelltest Selbsttest - Nase...   

                 Category      Sub_Category     Functional_Use  \
0  Health & Personal Care   Medical Product  Health Protection   
1     Electronics & Photo    Charging Cable    Charging / Data   
2     Electronics & Photo             Cable   Video Connection   
3     Electronics & Photo  Travel Accessory      Power Adapter   
4  Health & Personal Care   Medical Product       Medical Test   

      Physical_Property Age_Group DIY_Product Price_Range  
0               Textile       All          No         N/A  
1  Electronic Component       All          No         N/A  
2  Electronic Component       All          No         N/A  
3       Plastic / Metal     Adult          No         N/A  
4              Test Kit       All          No         N/A

Shape of Products DataFrame: (246, 9)


## Fuzzy matching implementation

### Subtask:
Implement a fuzzy matching function to compare product names in German, considering potential variations and typos.


**Reasoning**:
Install the required libraries for fuzzy matching.



In [6]:
!pip install fuzzywuzzy python-Levenshtein

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 159.9/159.9 kB 5.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 3.2/3.2 MB 61.8 MB/s eta 0:00:00


**Reasoning**:
Define the fuzzy matching function using the installed libraries.



In [7]:
from fuzzywuzzy import fuzz

def german_fuzzy_match(name1, name2):
  """
  Calculates the fuzzy similarity score between two German product names.

  Args:
    name1: The first product name (string).
    name2: The second product name (string).

  Returns:
    An integer representing the similarity score (0 to 100).
  """
  return fuzz.token_sort_ratio(name1, name2)

# Example usage (optional, for testing)
# score = german_fuzzy_match("Apple iPhone 13", "iPhone 13 Apple")
# print(f"Similarity score: {score}")

## Apply fuzzy matching and join

### Subtask:
Apply the fuzzy matching function to find the best matches between the two DataFrames and perform the join based on these matches.


**Reasoning**:
Apply the fuzzy matching function to find the best matches between the two DataFrames and perform the join based on these matches.



In [8]:
def find_best_match(product_name, products_df, threshold=80):
    """
    Finds the best matching Product_NUM from products_df for a given product name.

    Args:
        product_name: The product name from orders_df (string).
        products_df: The DataFrame containing product information (pandas DataFrame).
        threshold: The minimum fuzzy matching score to consider a match (integer, 0-100).

    Returns:
        The Product_NUM of the best match, or None if no match is found above the threshold.
    """
    best_score = -1
    best_match_product_num = None

    for index, row in products_df.iterrows():
        score = german_fuzzy_match(product_name, row['Product_Name'])
        if score > best_score and score >= threshold:
            best_score = score
            best_match_product_num = row['Product_NUM']

    return best_match_product_num

# Apply the function to orders_df to create a new column with matching Product_NUMs
orders_df['matched_Product_NUM'] = orders_df['productName'].apply(lambda x: find_best_match(x, products_df))

# Perform a left join of orders_df and products_df
joined_df = pd.merge(orders_df, products_df, left_on='matched_Product_NUM', right_on='Product_NUM', how='left')

# Display the head of the resulting joined DataFrame
display(joined_df.head())

eventDate                  orderDate marketplace  \
0  2025-07-17 14:36:06+00:00  2025-07-17 14:36:06+00:00          DE   
1  2025-07-08 05:06:46+00:00  2025-07-08 05:06:46+00:00          DE   
2  2025-07-04 05:36:24+00:00  2025-07-04 05:36:24+00:00          DE   
3  2025-07-02 19:54:58+00:00  2025-07-02 19:54:58+00:00          DE   
4  2025-06-17 17:33:15+00:00  2025-06-17 17:33:15+00:00          DE   

                                         productName    website  \
0  MELLERUD Schimmel Vernichter Aktivchlor | 1 x ...  Amazon.de   
1  Carson MicroFlip 100x-250x Taschenmikroskop mi...  Amazon.de   
2  Gosknor 2stk Octopus-Kopfhautmassagegerät, Har...  Amazon.de   
3      AJINOMOTO - Monosodium Glutamat, (1 X 200 GR)  Amazon.de   
4  Rainbow Designs Harry Potter Plüsch-Babyrassel...  Amazon.de   

               orderId currencyCode  quantity productCondition  totalOwed  \
0  028-1498008-9740330          EUR         1              New       7.99   
1  028-4603470-0628322          EUR         1              New      24.12   
2  028-5613038-3064361          EUR         1              New       9.59   
3  028-1018659-6139555          EUR         1              New       3.69   
4  028-4961403-1482742          EUR         1              New      14.05   

   ... matched_Product_NUM Product_NUM  \
0  ...              P00147      P00147   
1  ...              P00056      P00056   
2  ...              P00097      P00097   
3  ...              P00013      P00013   
4  ...              P00188      P00188   

                                        Product_Name                Category  \
0  MELLERUD Schimmel Vernichter Aktivchlor | 1 x ...  Health & Personal Care   
1  Carson MicroFlip 100x-250x Taschenmikroskop mi...     Electronics & Photo   
2  Gosknor 2stk Octopus-Kopfhautmassagegerät, Ha...  Health & Personal Care   
3      AJINOMOTO - Monosodium Glutamat, (1 X 200 GR)     Grocery & Beverages   
4  Rainbow Designs Harry Potter Plüsch-Babyrasse...                    Toys   

          Sub_Category               Functional_Use    Physical_Property  \
0   Household Cleaning  Mold Removal / Disinfection               Liquid   
1  Optics / Microscope    Magnification / Education     Optics / Plastic   
2             Massager         Massage / Relaxation      Resin / Plastic   
3   Cooking Ingredient           Flavor Enhancement  Solid / Crystalline   
4             Baby Toy               Sensory / Play       Fabric / Plush   

  Age_Group DIY_Product Price_Range  
0     Adult         Yes         N/A  
1       All          No         N/A  
2     Adult          No         N/A  
3       All          No         N/A  
4      Baby          No         N/A  

[5 rows x 25 columns]

## Create new table

### Subtask:
Write the resulting joined DataFrame to a new table named `Orders_Product_Enhanced` in the database.


**Reasoning**:
Write the resulting joined DataFrame to a new table named Orders_Product_Enhanced in the database.



In [9]:
import sqlite3

db_path = '/content/Categorized_DB_Enhanced (2).db'
table_name = 'Orders_Product_Enhanced'

try:
    conn = sqlite3.connect(db_path)
    # Write the DataFrame to a new table in the database
    joined_df.to_sql(table_name, conn, if_exists='replace', index=False)

    print(f"DataFrame successfully written to table '{table_name}' in the database.")

except sqlite3.Error as e:
    print(f"Error writing DataFrame to database: {e}")

finally:
    if conn:
        conn.close()

DataFrame successfully written to table 'Orders_Product_Enhanced' in the database.


## Verify new table

### Subtask:
List the tables in the database to confirm the creation of the new table.


**Reasoning**:
Connect to the database and list all tables to confirm the creation of the new table.



In [10]:
import sqlite3

db_path = '/content/Categorized_DB_Enhanced (2).db'

try:
    conn = sqlite3.connect(db_path)
    cursor = conn.cursor()

    cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
    tables = cursor.fetchall()

    print("Tables in the database:")
    for table in tables:
        print(table[0])

except sqlite3.Error as e:
    print(f"Error connecting to the database: {e}")

finally:
    if conn:
        conn.close()

Tables in the database:
order_history
Category_enhanced_orders
Demographics
PRODUCT_CATEGORY
Orders_Product_Enhanced


## Summary:

### Data Analysis Key Findings

*   Two tables, `Category_enhanced_orders` (266 rows, 15 columns) and `PRODUCT_CATEGORY` (246 rows, 9 columns), were successfully read from the database into pandas DataFrames.
*   A fuzzy matching function `german_fuzzy_match` using `fuzzywuzzy.fuzz.token_sort_ratio` was implemented to compare German product names.
*   A threshold of 80 was used to identify the best product name matches between the two tables.
*   A new column `matched_Product_NUM` was added to the `orders_df` DataFrame, containing the `Product_NUM` of the best fuzzy match from `products_df`.
*   A left join was performed between `orders_df` and `products_df` based on the fuzzy matching results, creating the `joined_df` DataFrame.
*   The resulting `joined_df` DataFrame was successfully written to a new table named `Orders_Product_Enhanced` in the database, replacing any existing table with the same name.
*   Verification confirmed the creation of the `Orders_Product_Enhanced` table in the database.

### Insights or Next Steps

*   Review the fuzzy matching threshold (currently 80) to ensure an optimal balance between precision and recall for the product name matches.
*   Analyze the `Orders_Product_Enhanced` table to understand the distribution of matched and unmatched orders and products.
